File: 프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
    answer = [0 for x in range(len(prices))]
    
    stack = [[0, prices[0]]]
    
    time = 1
    while time < len(prices):
        while stack and prices[time] < stack[-1][1]:
            cur = stack.pop()
            answer[cur[0]] = time - cur[0]
        stack.append([time, prices[time]])
        time += 1
    
    while stack:
        cur = stack.pop()
        answer[cur[0]] = (len(prices) - 1) - cur[0]
    
    return answer
    
    
    
        
        

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

public class Solution {
    public int[] solution(int[] prices) {
        var answer = new int[prices.length];
        var deque = new ArrayDeque<int[]>();
        
        deque.push(new int[]{0, prices[0]});
        
        int time = 1;
        while (time < prices.length) {
            while (!deque.isEmpty() && prices[time] < deque.peek()[1]) {
                int[] cur = deque.pop();
                answer[cur[0]] = time - cur[0];
            }
            deque.push(new int[]{time, prices[time]});
            time++;
        }
        
        while (!deque.isEmpty()) {
            int[] cur = deque.pop();
            answer[cur[0]] = (prices.length - 1) - cur[0];
        }
        
        return answer;
    }
}
